File: VB/Northwind.vb
Imports Microsoft.VisualBasic
	Imports System
	Imports DevExpress.Xpo
Namespace Northwind

	<Persistent("Customers")> _
	Public Class Customer
		Inherits XPLiteObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _customerID As String

		<Key> _
		Public Property CustomerID() As String
			Get
				Return _customerID
			End Get
			Set(ByVal value As String)
				SetPropertyValue("CustomerID", _customerID, value)
			End Set
		End Property

		Private _contactTitle As String
		Public Property ContactTitle() As String
			Get
				Return _contactTitle
			End Get
			Set(ByVal value As String)
				SetPropertyValue("ContactTitle", _contactTitle, value)
			End Set
		End Property

		Private _companyName As String
		Public Property CompanyName() As String
			Get
				Return _companyName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("CompanyName", _companyName, value)
			End Set
		End Property

	Private _address As String
		Public Property Address() As String
			Get
				Return _address
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Address", _address, value)
			End Set
		End Property

		<Association("CustomerOrders", GetType(Order))> _
		Public ReadOnly Property Orders() As XPCollection(Of Order)
			Get
				Return GetCollection(Of Order)("Orders")
			End Get
		End Property
	End Class

	<Persistent("Orders")> _
	Public Class Order
		Inherits XPLiteObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		<Key(True), Persistent> _
		Private OrderID As Integer = -1

		<PersistentAlias("OrderID")> _
		Public ReadOnly Property ID() As Integer
			Get
				Return OrderID
			End Get
		End Property

		Private _OrderDate As DateTime
		Public Property OrderDate() As DateTime
			Get
				Return _OrderDate
			End Get
			Set(ByVal value As DateTime)
				SetPropertyValue("OrderDate", _OrderDate, value)
			End Set
		End Property

		<PersistentAlias("Custom('GetMonth', [OrderDate])")> _
		Public ReadOnly Property OrderMonth() As Integer
			Get
				' .NET implementation of the custom function (GetMonth)
				Return OrderDate.Month
			End Get
		End Property

		Private _customer As Customer
		<Persistent("CustomerID"), Association("CustomerOrders")> _
		Public Property Customer() As Customer
			Get
				Return _customer
			End Get
			Set(ByVal value As Customer)
				SetPropertyValue("Customer", _customer, value)
			End Set
		End Property

		Private _Employer As Employee
		<Persistent("EmployeeID")> _
		Public Property Employee() As Employee
			Get
				Return _Employer
			End Get
			Set(ByVal value As Employee)
				SetPropertyValue("Employer", _Employer, value)
			End Set
		End Property
	End Class

	<Persistent("Employees")> _
	Public Class Employee
		Inherits XPLiteObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		<Key(True), Persistent> _
		Public EmployeeID As Integer

		Private _FirstName As String
		Public Property FirstName() As String
			Get
				Return _FirstName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("FirstName", _FirstName, value)
			End Set
		End Property
		Private _LastName As String
		Public Property LastName() As String
			Get
				Return _LastName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("LastName", _LastName, value)
			End Set
		End Property

		<PersistentAlias("Concat([FirstName], ' ', [LastName])")> _
		Public ReadOnly Property FullName() As String
			Get
				Return CStr(EvaluateAlias("FullName"))
			End Get
		End Property
	End Class
End Namespace

File: CS/Northwind.cs
namespace Northwind {
    using System;
    using DevExpress.Xpo;

    [Persistent("Customers")]
    public class Customer : XPLiteObject {
        public Customer(Session session) : base(session) { }

        string _customerID;
        
        [Key]
        public string CustomerID {
            get { return _customerID; }
            set { SetPropertyValue("CustomerID", ref _customerID, value); }
        }

        string _contactTitle;
        public string ContactTitle {
            get { return _contactTitle; }
            set { SetPropertyValue("ContactTitle", ref _contactTitle, value); }
        }

        string _companyName;
        public string CompanyName {
            get { return _companyName; }
            set { SetPropertyValue("CompanyName", ref _companyName, value); }
        }

	string _address;
        public string Address {
            get { return _address; }
            set { SetPropertyValue("Address", ref _address, value); }
        }        

        [Association("CustomerOrders", typeof(Order))]
        public XPCollection<Order> Orders {
            get {
                return GetCollection<Order>("Orders");
            }
        }
    }

    [Persistent("Orders")]
    public class Order : XPLiteObject {
        public Order(Session session) : base(session) { }

        [Key(true), Persistent]
        int OrderID = -1;

        [PersistentAlias("OrderID")]
        public int ID {
            get { return OrderID; }
        }

        private DateTime _OrderDate;
        public DateTime OrderDate {
            get {
                return _OrderDate;
            }
            set {
                SetPropertyValue("OrderDate", ref _OrderDate, value);
            }
        }

        [PersistentAlias("Custom('GetMonth', [OrderDate])")]
        public int OrderMonth {
            get {
                // .NET implementation of the custom function (GetMonth)
                return OrderDate.Month;
            }
        }

        Customer _customer;
        [Persistent("CustomerID"), Association("CustomerOrders")]
        public Customer Customer {
            get { return _customer; }
            set { SetPropertyValue("Customer", ref _customer, value); }
        }

        private Employee _Employer;
        [Persistent("EmployeeID")]
        public Employee Employee {
            get {
                return _Employer;
            }
            set {
                SetPropertyValue("Employer", ref _Employer, value);
            }
        }
    }

    [Persistent("Employees")]
    public class Employee : XPLiteObject {
        public Employee(Session session) : base(session) { }

        [Key(true), Persistent]
        public int EmployeeID;

        private string _FirstName;
        public string FirstName {
            get {
                return _FirstName;
            }
            set {
                SetPropertyValue("FirstName", ref _FirstName, value);
            }
        }
        private string _LastName;
        public string LastName {
            get {
                return _LastName;
            }
            set {
                SetPropertyValue("LastName", ref _LastName, value);
            }
        }

        [PersistentAlias("Concat([FirstName], ' ', [LastName])")]
        public string FullName {
            get {
                return (string)EvaluateAlias("FullName");
            }
        }
    }
}
